docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,319 @@
+
+Since we're scanning meta2, the respective node has an opportunity to update
+its range descriptor cache. Which suggests that perhaps all nodes should do
+this.
+
+The resulting report will be saved in storage under a non-versioned key. The
+two versions of the report will be stored at all times.
+
+### Collecting lease and size information
+
+The current leaseholder is not reflected in the range descriptor (i.e. in
+meta2). So, in order to report on leaseholder preference conformance, the node
+generating the report needs to collect this information in another way. We'll
+add a new RPC asking a node to return all the leases for its replicas
+(including info on ranges with no lease). The node generating the report will
+ask all other nodes for their leases and join that information with meta2.
+This RPC will also return size information on all the ranges, which sizes will
+be summed up in the conformance report for the ranges with the same problem.
+
+Since lease information is gathered distinctly from range information, the two
+might not be consistent: ranges could have appeared and dissapeared in between.
+To keep it simple, we'll consider the meta2 the source of truth, and we'll
+ignore lease information when it doesn't match with that.
+
+Multiple nodes are expected to return information on the same replica; we'll
+consider the latest lease among them.
+If multiple nodes fails to respond and so we don't have info on some ranges,
+for the purposes of the report we'll consider all the respective ranges to not
+have an active lease (and presumably different alerting will fire since the
+unresponsive nodes are unhealthy).
+
+
+The implementation of the view `crdb_internal.ranges` will also change to take
+advantage of this new RPC. Currently, it's a view on top of the virtual table
+`crdb_internal.ranges_no_leases` executing an
+`crdb_internal.lease_holder(start_key)` (i.e. a `LeaseInfo` request) for every
+range. That's too expensive. Once we moved to the new mechanism, we can also
+deprecate `crdb_internal.ranges_no_leases`.
+
+As an incidental benefit of collecting lease information this way, the node
+collecting all this information can update its leaseholder cache. Which
+suggests that perhaps all nodes should exchange this info with each other.
+
+### Notes
+
+The notion of a node being alive or dead will be more nuanced in this reporting
+code than it is in the allocator. The allocator only concerns itself with nodes
+that have been dead for 5 minutes (cluster setting server.time_until_store_dead)
+- that’s when it starts moving replicas away. Nodes that died more recently are
+as good as live ones. But not so for this report; we can’t claim that everything
+is fine for 5 minutes after a node’s death. For the purposes of reporting
+under-replication, constraint and lease preference conformance, a node
+will be considered dead a few seconds after it failed to ping its liveness
+record. Replicas on dead nodes are discarded.
```

Comment:
We will probably need to note something about waiting 5 minutes before taking action in the webui in case it recovers. its more about a don't do further harm kind of idea

Reply:
like saying "1GB underreplicated. Will start repairing in 5min." ?

Reply:
yeah i like that idea.